pagination/pagination.go

```diff
@@ -0,0 +1,129 @@
+// that is later used in the json encoding to populate the hrefs dynamiclly
+// some strings are pointers so they can be a null JSON value if empty
+type Pagination struct {
+	PageNumber        int     `json:"page_number"`
+	PagesTotal        int     `json:"pages_total"`
+	ItemsPerPage      int     `json:"items_per_page"`
+	ItemsPerPageLimit int     `json:"items_per_page_limit"`
+	ItemsTotal        int     `json:"items_total"`
+	FirstHref         string  `json:"first_href"`
+	LastHref          string  `json:"last_href"`
+	NextHref          *string `json:"next_href"`
+	PrevHref          *string `json:"prev_href"`
+	request           *http.Request
+}
+
+// GetOffset returns the current offset (i.e. for database queries)
+// based on the current page and number of itmes
+func (p *Pagination) GetOffset() int {
+	if p.PageNumber == 1 {
+		return 0
+	}
+	return (p.PageNumber - 1) * p.ItemsPerPage
+}
+
+// SetItemsTotal will set the ItemsTotal value as well as the dynamic PagesTotal based on ItemsPerPage
+func (p *Pagination) SetItemsTotal(i int) {
+	p.ItemsTotal = i
+	p.PagesTotal = int(math.Ceil(float64(i) / float64(p.ItemsPerPage)))
+}
+
+// getURL is a helper used to build the HREFs based on the current URL
+// it really just overrides the currently set page query string with the required value
+func (p *Pagination) getURL(page int) (u *url.URL) {
+	u = p.request.URL
+
+	q := u.Query()
+	q.Set("page", strconv.Itoa(page))
+	q.Set("limit", strconv.Itoa(p.ItemsPerPage))
+	u.RawQuery = q.Encode()
+
+	// The URL does not always contain the required information, so if its not there
+	// build it from the request info
+	if "" == u.Host {
+		u.Host = p.request.Host
+	}
+	if "" == u.Scheme {
+		u.Scheme = "http"
+		if nil != p.request.TLS {
+			u.Scheme = "https"
+		}
+	}
+
+	return
+}
+
+// New returns a configured Pagination struct that can be used
+func New(PageNumber int, ItemsPerPage int, ItemsPerPageLimit int, r *http.Request) (p Pagination, err error) {
+	if ItemsPerPage > ItemsPerPageLimit {
+		return p, fmt.Errorf("The requested number of items per page (%d) is greater than the maximum allowed (%d)", ItemsPerPage, ItemsPerPageLimit)
+	}
+
+	if PageNumber < 0 {
+		return p, fmt.Errorf("The requested page (%d) is less than 1", PageNumber)
+	}
+
+	if ItemsPerPage < 0 {
+		return p, fmt.Errorf("The requested items per page (%d) is less than 1", ItemsPerPage)
+	}
+
+	// defaults
+	if 0 == PageNumber {
+		PageNumber = defaultPageNumber
+	}
+
+	if 0 == ItemsPerPage {
+		ItemsPerPage = defaultItemsPerPage
+	}
+
+	return Pagination{
+		PageNumber:        PageNumber,
+		ItemsPerPage:      ItemsPerPage,
+		ItemsPerPageLimit: ItemsPerPageLimit,
+		request:           r,
+	}, err
+}
+
+// MarshalJSON is called when ever the Pagination object is json encoded, we
+// use this chance to set the correct href values (as they depend on other data being set, like number of results)
+func (p Pagination) MarshalJSON() ([]byte, error) {
```

Comment:
- Calling `MarshalJSON` without calling `SetTotal` will result in returning 0 for the total... (nothing denotes that the total is unknown, the same for PagesTotal)
- When you call `MarshallJSON` without setting the total the `last_href` will return `url?page=0` and the `first_page` will be `url?page=1` (weird)